File: hanamikoji/go-backend/agent.go
package main

import "math/rand"

type Agent interface {
	ChooseAction(obs Observation, validActions []ActionName) (ActionName, error)
	ChooseCards(obs Observation, n int) ([]Suit, error)
	Respond(obs Observation) (int, error)
}

type RandomAgent struct{}

func (r *RandomAgent) ChooseAction(_ Observation, valid []ActionName) (ActionName, error) {
	return valid[rand.Intn(len(valid))], nil
}

func (r *RandomAgent) ChooseCards(obs Observation, n int) ([]Suit, error) {
	hand := append([]Suit{}, obs.MyHand...)
	rand.Shuffle(len(hand), func(i, j int) { hand[i], hand[j] = hand[j], hand[i] })
	return hand[:n], nil
}

func (r *RandomAgent) Respond(obs Observation) (int, error) {
	if len(obs.OfferedGift) > 0 {
		return rand.Intn(len(obs.OfferedGift)), nil
	}
	return rand.Intn(2), nil
}

File: hanamikoji/python-ml/agent.py
import sys
import json


def choose_action(obs: dict, valid_actions: list[str]) -> str:
    # stub: pick first valid action
    return valid_actions[0]


def choose_cards(obs: dict, n: int) -> list[str]:
    # stub: pick first n cards from hand
    return obs["MyHand"][:n]


def respond(obs: dict) -> int:
    # stub: always pick option 0
    return 0


def decide(req: dict) -> object:
    step = req["step"]
    obs = req["observation"]

    if step == "ChooseAction":
        return choose_action(obs, req["valid_actions"])
    elif step == "ChooseCards":
        return choose_cards(obs, req["n_cards_to_choose"])
    elif step == "OpponentResponse":
        return respond(obs)
    else:
        raise ValueError(f"unknown step: {step}")


def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        req = json.loads(line)
        result = decide(req)
        print(json.dumps(result), flush=True)


if __name__ == "__main__":
    main()
